### DISCUSSION TITLE
Initialize async API

### DISCUSSION THREAD

**Discussion by herecydev** [remix-run/remix#3341] [Proposals]
Support an API "hook" that runs very early when loading the `remix.config.js`. You can currently hack it with routes as [that supports async](https://github.com/remix-run/remix/blob/main/packages/remix-dev/config.ts#L391) but would be better to support a more idiomatic api.

Example:

```js
module.exports = {
  serverBuildTarget: "cloudflare-pages",
  server: "./server.js",
  initialize: async () => {
    // do some async work here that might be needed before building
  },
  pages: async () => {
  }
};
```

Motivation: I need to create localized text labels that would be generated at "build time", so the same app could be translated in English, Spanish, etc. This would be done within the `initialize` function, fetching the translations, building a file and then the app code would reference that generated file

**Comment by kiliman** [comment]
Here’s a tip. When you are suggesting changes it’s always helpful to include your usecase. Ryan responds better to actual needs vs “what-ifs”. It helps them to come up with a good design vs just haphazardly adding extensibility points.

**Reply by herecydev** [reply]
Updated with a motivation

**Comment by sergiodxa** [comment]
How do you want to access the data initialized there?

**Reply by herecydev** [reply]
```js
const fs = require("fs-extra")

module.exports = {
  ignoredRouteFiles: ["**/.*"],
  initialize: () => {
    // go and fetch some data from 3rd party, database, etc
    await fs.outputFile('./labels.ts', 'export default { buttonText: "Buy" }')
  }
};
```

And then referenced in app:

```js
import labels from "labels";

export default function Index() {
  return <button>{labels.buttonText}</button>
}

```

**Reply by sergiodxa** [reply]
So you want to generate a module and be able to import it, why not run a prebuild script then?
You could create a `scripts/labels.js` file in your project (outside app) and in the package.json add

```json
{
  "scripts": {
    "prebuild": "node scripts/labels.js",
    "build": "remix build",
    "dev": "remix dev",
    "start": "remix-serve build"
  }
}
```

Then, every time you run `npm run build` it will first run your scripts, generate that `labels.ts` file and your app can import it.

**Reply by herecydev** [reply]
On the surface that might be fine, is it enough to rely on all adapters supporting running multiple scripts? I don't have enough experience in every adapter. For example if an adapter can only run a single command like `remix build`?

**Reply by sergiodxa** [reply]
If the adapter runs `npm run build` you can rely on `prebuild`, and I'm sure they all do it.

**Reply by herecydev** [reply]
Fine; it looks like it will solve the use case I have so I appreciate you reaching out with a solution.
